src/mds/Server.cc

```diff
@@ -1118,6 +1118,14 @@ void Server::find_idle_sessions()
 	     << "), not marking any client stale" << dendl;
     return;
   }
+  // don't kick client if ops or osds are laggy
+  if(mds->objecter->get_num_osd_laggy()) {
```

Comment:
Unfortunately, we can't use the objecter's builtins for this. It only reports laggy osds for which the MDS has a session open with AND has laggy ops with.

Consider the case of a secondary data pool on e.g. hdd devices. The MDS will not typically open sessions with the OSD unless deleting data. So, if a client is experiencing slow ops with those OSDs, the MDS may not know that (via this objecter method).

I think the robust solution here is either:

- Check if _any_ osd is laggy (as reported by the osdmap)
- (Slightly more robust:) check if any OSD which has pgs for any meta or data pool is laggy
- Or: have clients report laggy osd ops in a regular heartbeat to the mds (requires clients changes so not great)

Reply:
thanks for the review patrick, I'll go through the code and push the patch, will ping you once ready

Reply:
Discussed this briefly with @dparmar18 and I don't think it's worth getting more complicated than "is an OSD laggy". Mapping laggy OSDs to their PGs is not completely trivial, and a single laggy OSD can cause back pressure issues to slow down other OSDs anyway.

Reply:
I agree with @gregsfortytwo, looking at an approach to find a laggy OSD then find the PGs (which are part of meta/data pool) mapped to it looks clumsy and would incur significant time complexity. It's doable to but it would be too slow.

Reply:
> Discussed this briefly with @dparmar18 and I don't think it's worth getting more complicated than "is an OSD laggy". Mapping laggy OSDs to their PGs is not completely trivial, and a single laggy OSD can cause back pressure issues to slow down other OSDs anyway.

+1 for the explanation and replying on a straightforward check.

Reply:
moving forward with this approach, thanks guys